src/types/string.rs

```diff
@@ -242,6 +242,7 @@ impl PyString {
         #[cfg(not(any(Py_3_10, not(Py_LIMITED_API))))]
         {
             let bytes = unsafe {
+                #[allow(deprecated)]
                 self.py()
                     .from_owned_ptr_or_err::<PyBytes>(ffi::PyUnicode_AsUTF8String(self.as_ptr()))
```

Comment:
I couldn't work out how to fix this one. The main problem I had was trying to return a reference to owned data going out of scope.

Reply:
I think something like might work
```suggestion
ffi::PyUnicode_AsUTF8String(self.as_ptr()).assume_owned_or_err(self.py()).downcast_into_unchecked::<PyBytes>()
```

Reply:
```rust
error[E0308]: mismatched types
   --> src/types/string.rs:244:13
    |
236 |       pub fn to_str(&self) -> PyResult<&str> {
    |                               -------------- expected `Result<&str, err::PyErr>` because of return type
...
244 | /             ffi::PyUnicode_AsUTF8String(self.as_ptr())
245 | |                 .assume_owned_or_err(self.py())
246 | |                 .downcast_into_unchecked::<PyBytes>()
    | |_____________________________________________________^ expected `Result<&str, PyErr>`, found `Result<Bound<'_, ...>, ...>`
    |
    = note: expected enum `Result<&str, _>`
               found enum `Result<instance::Bound<'_, bytes::PyBytes>, _>`
```

Reply:
This one is a fundamental part of the GIL Ref API; there's no way around it. The problem is that this function creates a new `bytes` object and then returns data from it. That worked with the GIL Refs where everything went in the pool, but it makes no sense in an API where things follow normal ownership rules.

The best you can do here is probably

```suggestion
                let bytes = self.as_borrowed().encode_utf8()?.into_gil_ref();
                Ok(unsafe { std::str::from_utf8_unchecked(bytes.as_bytes()) })
```

Reply:
What are we going to do when we remove `into_gil_ref`? Do we need to think about that yet?

Reply:
For 0.21 this solution will work with an `#[allow(deprecated)]`, and then in 0.22 if the `gil-refs` feature is disabled this function will be gone too.

(The `PyStringMethods`  implentation of `to_str` has a Python 3.10 restriction and has this logic of making bytes first in `to_cow`, which has to copy the bytes into a new `String`.)